ports/qemu-riscv/Makefile

```diff
@@ -0,0 +1,127 @@
+include ../../py/mkenv.mk
+-include mpconfigport.mk
+
+# qstr definitions (must come before including py.mk)
+QSTR_DEFS = qstrdefsport.h
+
+# MicroPython feature configurations
+MICROPY_ROM_TEXT_COMPRESSION ?= 1
+
+# include py core make definitions
+include $(TOP)/py/py.mk
+include $(TOP)/extmod/extmod.mk
+
+BOARD ?= virt
+
+CROSS_COMPILE ?= riscv64-unknown-elf-
+
+GCC_VERSION = $(word 1, $(subst ., , $(shell $(CC) -dumpversion)))
+
+# Ubuntu 22.04 ships with Picolibc rather than newlib.  To avoid having
+# per-distro workarounds, and given that Picolibc may get more support
+# in the future, it is chosen over newlib if it is found.
+PICOLIBC_SPECS = $(shell $(CC) --print-file-name=picolibc.specs)
+ifeq ($(PICOLIBC_SPECS),picolibc.specs)
+# Picolibc was not found.
+SPECS_FRAGMENT =
+else
+SPECS_FRAGMENT = --specs=$(PICOLIBC_SPECS)
+CFLAGS += $(SPECS_FRAGMENT)
+endif
+
+ifeq ($(BOARD),virt)
+ABI = ilp32
+# GCC 10 and lower do not recognise the Zicsr extension in the
+# architecture name.  "Make" unfortunately does not provide any simple way
+# to perform numeric comparisons, so to keep things simple we assume that
+# GCC is at least version 10 for the time being.
+ifeq ($(GCC_VERSION),10)
+ARCH ?= rv32imac
+else
+# Recent GCC versions explicitly require to declare extensions.
+ARCH ?= rv32imac_zicsr
+ARCH_LD ?= rv32imac_zicsr
+endif
+AFLAGS = -mabi=$(ABI) -march=$(ARCH)
+CFLAGS += $(AFLAGS)
+CFLAGS += -DQEMU_SOC_VIRT
+ARCH_LD ?= $(ARCH)
+LDSCRIPT = virt.ld
+LDFLAGS += -mabi=$(ABI) -march=$(ARCH_LD) -T $(LDSCRIPT) -Wl,-EL
+SRC_BOARD_O = shared/runtime/gchelper_native.o shared/runtime/gchelper_rv32i.o
+SRC_BOARD_O += entrypoint.o
+endif
+
+INC += -I.
+INC += -I$(TOP)
+INC += -I$(BUILD)
+
+CFLAGS += $(INC) -Wall -Wpointer-arith -Wdouble-promotion -Wfloat-conversion -Werror -std=gnu99 $(COPT) \
+	 -ffunction-sections -fdata-sections
+CFLAGS += $(CFLAGS_EXTRA)
+
+LD = $(CC)
+
+# Debugging/Optimization
+ifeq ($(DEBUG), 1)
+CFLAGS += -g
+COPT = -O0
+else
+COPT += -Os -DNDEBUG
+endif
+
+LDFLAGS += $(SPECS_FRAGMENT) -Wl,--gc-sections -Wl,-Map=$(@:.elf=.map)
+
+SRC_COMMON_C = \
+	interrupts.c \
+	startup.c \
+	uart.c \
+	shared/libc/string0.c \
+	shared/runtime/sys_stdio_mphal.c \
+
+SRC_RUN_C = \
+	../qemu-arm/main.c \
+
+SRC_TEST_C = \
+	test_main.c \
+	lib/tinytest/tinytest.c \
+
+LIB_SRC_C += $(SRC_LIB_LIBM_C)
+LIB_SRC_C += $(SRC_LIB_LIBM_SQRT_SW_C)
+
+OBJ_COMMON =
+OBJ_COMMON += $(PY_O)
+OBJ_COMMON += $(addprefix $(BUILD)/, $(SRC_COMMON_C:.c=.o))
+OBJ_COMMON += $(addprefix $(BUILD)/, $(SRC_BOARD_O))
+OBJ_COMMON += $(addprefix $(BUILD)/, $(LIB_SRC_C:.c=.o))
+
+OBJ_RUN =
+OBJ_RUN += $(addprefix $(BUILD)/, $(SRC_RUN_C:.c=.o))
```

Comment:
While testing this with an ubuntu container I noticed that this builds `qemu-arm/main.o` in the port directory rather than `$(BUILD)/qemu-arm/main.o`.

I did find one possible workaround, but I'm not too impressed by it. Damien might know a better way, I don't know if this pattern exists in other ports.

```diff
diff --git i/ports/qemu-riscv/Makefile w/ports/qemu-riscv/Makefile
index 35445cd48e..00eaccca3a 100644
--- i/ports/qemu-riscv/Makefile
+++ w/ports/qemu-riscv/Makefile
@@ -80,7 +80,7 @@ SRC_COMMON_C = \
        shared/runtime/sys_stdio_mphal.c \
 
 SRC_RUN_C = \
-       ../qemu-arm/main.c \
+       $(TOP)/ports/qemu-arm/main.c \
 
 SRC_TEST_C = \
        test_main.c \
@@ -96,7 +96,7 @@ OBJ_COMMON += $(addprefix $(BUILD)/, $(SRC_BOARD_O))
 OBJ_COMMON += $(addprefix $(BUILD)/, $(LIB_SRC_C:.c=.o))
 
 OBJ_RUN =
-OBJ_RUN += $(addprefix $(BUILD)/, $(SRC_RUN_C:.c=.o))
+OBJ_RUN += $(subst $(TOP),$(BUILD),$(SRC_RUN_C:.c=.o))
 
 ALL_OBJ_RUN = $(OBJ_COMMON) $(OBJ_RUN)
```

Reply:
I guess the less "clever" version is `OBJ_RUN += $(BUILD)/ports/qemu-arm/main.o`

Reply:
The simplest (?) way to do this, is how other ports and files from, eg, `lib` and `shared` work: just specify the source relative to top:
```make
SRC_RUN_C = ports/qemu-arm/main.c
```

Reply:
Oh yeah! I don't know why I thought this wouldn't work

Reply:
The current version should keep object files in the same `build` directory.